### PR TITLE
kvserver: fix replicate queue metrics tracking bug

### LINKED ISSUE #86107: roachtest: replicate/wide failed
roachtest.replicate/wide [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6087204?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6087204?buildTab=artifacts#/replicate/wide) on master @ [d25cb57ccd9bc643ce9058ebd2057cab36b69ad5](https://github.com/cockroachdb/cockroach/commits/d25cb57ccd9bc643ce9058ebd2057cab36b69ad5):

```
test artifacts and logs in: /artifacts/replicate/wide/run_1
	allocator.go:381,allocator.go:388,allocator.go:459,test_runner.go:896: dial tcp 34.73.248.131:26257: connect: connection refused
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=1</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*replicate/wide.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-18577

### LINKED ISSUE #86374: roachtest: loqrecovery/workload=movr/rangeSize=2mb failed
roachtest.loqrecovery/workload=movr/rangeSize=2mb [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6137147?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6137147?buildTab=artifacts#/loqrecovery/workload=movr/rangeSize=2mb) on master @ [5c2c62ecf1bea60c807edc6b4da22d900ad4ae03](https://github.com/cockroachdb/cockroach/commits/5c2c62ecf1bea60c807edc6b4da22d900ad4ae03):

```
test artifacts and logs in: /artifacts/loqrecovery/workload=movr/rangeSize=2mb/run_1
	assertions.go:264,assertions.go:1392,require.go:1261,loss_of_quorum_recovery.go:296,monitor.go:105,errgroup.go:74: 
			Error Trace:	/go/src/github.com/cockroachdb/cockroach/loss_of_quorum_recovery.go:296
			            				/go/src/github.com/cockroachdb/cockroach/monitor.go:105
			            				/go/src/github.com/cockroachdb/cockroach/errgroup.go:74
			            				/go/src/github.com/cockroachdb/cockroach/asm_amd64.s:1571
			Error:      	Received unexpected error:
			            	dial tcp 34.75.34.5:26257: connect: connection refused
			Test:       	loqrecovery/workload=movr/rangeSize=2mb

	test_impl.go:265,require.go:1264,loss_of_quorum_recovery.go:296,monitor.go:105,errgroup.go:74: 
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*loqrecovery/workload=movr/rangeSize=2mb.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-18725

### LINKED ISSUE #86544: roachtest: loqrecovery/workload=movr/rangeSize=default failed
roachtest.loqrecovery/workload=movr/rangeSize=default [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6181118?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6181118?buildTab=artifacts#/loqrecovery/workload=movr/rangeSize=default) on master @ [aaf50e920ceff3c2863ab96b9e3614b8434b70a8](https://github.com/cockroachdb/cockroach/commits/aaf50e920ceff3c2863ab96b9e3614b8434b70a8):

```
test artifacts and logs in: /artifacts/loqrecovery/workload=movr/rangeSize=default/run_1
	assertions.go:264,assertions.go:1392,require.go:1261,loss_of_quorum_recovery.go:296,monitor.go:105,errgroup.go:74: 
			Error Trace:	/go/src/github.com/cockroachdb/cockroach/loss_of_quorum_recovery.go:296
			            				/go/src/github.com/cockroachdb/cockroach/monitor.go:105
			            				/go/src/github.com/cockroachdb/cockroach/errgroup.go:74
			            				/go/src/github.com/cockroachdb/cockroach/asm_amd64.s:1571
			Error:      	Received unexpected error:
			            	dial tcp 34.138.143.5:26257: connect: connection refused
			Test:       	loqrecovery/workload=movr/rangeSize=default

	test_impl.go:265,require.go:1264,loss_of_quorum_recovery.go:296,monitor.go:105,errgroup.go:74: 
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*loqrecovery/workload=movr/rangeSize=default.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-18806

### PR DESCRIPTION
**Pull request by AlexTalks**
In #85844, tracking replicate queue success and error metrics by
allocator action was introduced, however this change inadvertently
included a bug due to missing the `RemoveDead(Non)Voter` allocator
action case, in combination with a panic on unexpected actions. This
change fixes the missing case as well as the panic.

Fixes #86544, fixes #86374, fixes #86107.

Release justification: Bug fix
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replicate_queue.go [hunk 1]
```diff
@@ -481,6 +481,12 @@ func (metrics *ReplicateQueueMetrics) trackResultByAllocatorAction(
 		} else {
 			metrics.ReplaceDeadReplicaErrorCount.Inc(1)
 		}
+	case allocatorimpl.AllocatorRemoveDeadVoter, allocatorimpl.AllocatorRemoveDeadNonVoter:
```
Comment (irfansharif) [inline pkg/kv/kvserver/replicate_queue.go:484]: Should this case be folded into the one right above?

File: pkg/kv/kvserver/replicate_queue.go [hunk 2]
```diff
@@ -494,7 +500,7 @@ func (metrics *ReplicateQueueMetrics) trackResultByAllocatorAction(
 			metrics.RemoveDecommissioningReplicaErrorCount.Inc(1)
 		}
 	default:
-		panic(fmt.Sprintf("unsupported AllocatorAction: %v", action))
+		// Unsupported AllocatorAction.
```
Comment (irfansharif) [inline pkg/kv/kvserver/replicate_queue.go:503]: Worth log.Errorf-ing, or fatal-ing under `buildutil.CrdbTestBuild`.

**Review by irfansharif — APPROVED** [review]

**Comment by AlexTalks** [comment]
The reason this wasn't caught in tests was because `TestReplicateQueueDeadNonVoters` is skipped, per #76996.  I spent a bit of time attempting to unskip this, but haven't resolved that just yet.

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86844)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @irfansharif)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86844#-NAIw2RY0v8TCXnqy6Kc-r1-484:-NALR4oB2_YMnMHDsr0J:b-ussihm) ([raw file](https://github.com/cockroachdb/cockroach/blob/053abc0801c3d4fee0a47603d091c4b538f9eaab/pkg/kv/kvserver/replicate_queue.go#L484)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Should this case be folded into the one right above?
</blockquote></details>

No, we have separate metrics for `ReplaceDead` and `RemoveDead`.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86844#-NAIw2RY0v8TCXnqy6Kc-r1-503:-NALRJK531YICMFBIDNk:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/053abc0801c3d4fee0a47603d091c4b538f9eaab/pkg/kv/kvserver/replicate_queue.go#L503)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Worth log.Errorf-ing, or fatal-ing under `buildutil.CrdbTestBuild`.
</blockquote></details>

Will do.

**Review by AlexTalks — COMMENTED** [review]
> Dismissed @irfansharif from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86844)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86844#-NAIw2RY0v8TCXnqy6Kc-r1-503:-NALXL4TEr0sIGzaK5IX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/053abc0801c3d4fee0a47603d091c4b538f9eaab/pkg/kv/kvserver/replicate_queue.go#L503)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Will do.
</blockquote></details>

Done.

**Comment by AlexTalks** [comment]
bors cancel
(Just noticed this is going to have a merge conflict)

**Pull request merged**